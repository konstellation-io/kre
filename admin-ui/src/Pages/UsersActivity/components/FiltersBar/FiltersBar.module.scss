@import 'Styles/mixins';
@import 'Styles/colors';

.formField {
  height: auto;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  background-color: palette(lowlight, 900);
  padding: 11px 25px;

  .label {
    @include font-body;
    color: font-color(light);

    margin: 0 40px 0 30px;
  }

  > div:not(:last-child) {
    flex: 1 1 0px;
    min-width: 210px;
    margin-right: 2vw;
  }
}

.buttons {
  display: flex;
  flex-direction: row;
}

.customLabel {
  @include font-button;

  text-transform: uppercase;
  display: flex;

  span {
    flex-shrink: 0;
  }

  .customLabelHighlight {
    @include font-small;

    text-transform: capitalize;
    margin-left: $grid-unit;

    &.highlight {
      color: palette(highlight);
    }
    &.feedback {
      color: palette(feedback);
    }
    &.success {
      color: palette(success);
    }
    &.regular {
      color: font-color(regular);
    }
  }
}
