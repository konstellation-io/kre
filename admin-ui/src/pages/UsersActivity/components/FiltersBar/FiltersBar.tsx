import React, { useEffect } from 'react';

import { get } from 'lodash';
import Calendar from '../../../../components/Form/Calendar/Calendar';
import Select from '../../../../components/Form/Select/Select';
import Button from '../../../../components/Button/Button';
import SearchSelect from '../../../../components/Form/SearchSelect/SearchSelect';
import * as CHECK from '../../../../components/Form/check';

import styles from './FiltersBar.module.scss';
import { ApolloError } from 'apollo-client';
import { useForm } from 'react-hook-form';
import { Moment } from 'moment';

export const typeToText = {
  LOGIN: 'Login',
  LOGOUT: 'Logout',
  CREATE_RUNTIME: 'Create Runtime',
  CREATE_VERSION: 'Create Version',
  PUBLISH_VERSION: 'Publish Version',
  UNPUBLISH_VERSION: 'Unpublish Version',
  STOP_VERSION: 'Stop Version',
  START_VERSION: 'Start Version',
  UPDATE_SETTING: 'Update Settings',
  UPDATE_VERSION_CONFIGURATION: 'Update Version Configuration'
};

export type UserActivityFormData = {
  type: string;
  userEmail: string;
  fromDate: Moment;
  toDate: Moment;
};

type FormFieldProps = {
  error?: ApolloError;
  onSubmit: (form: UserActivityFormData) => void;
  types: string[];
  users: string[];
};

function FiltersBar({ onSubmit, types, users }: FormFieldProps) {
  const { handleSubmit, register, setValue, errors, watch, reset } = useForm<
    UserActivityFormData
  >({
    reValidateMode: 'onBlur'
  });
  useEffect(() => {
    register({ name: 'type' });
    register(
      { name: 'userEmail' },
      {
        validate: (value: string) => {
          return CHECK.getValidationError([
            CHECK.isFieldNotInList(
              value,
              users,
              true,
              'The user must be from the list'
            )
          ]);
        }
      }
    );
    register({ name: 'fromDate' });
    register({ name: 'toDate' });
  }, [users, register]);

  function handleOnSubmit(formData: UserActivityFormData) {
    onSubmit(formData);
  }

  return (
    <form className={styles.formField}>
      <Select
        options={types}
        onChange={(value: string) => setValue('type', value)}
        error={get(errors.type, 'message')}
        formSelectedOption={watch('type')}
        label="Activity type"
        placeholder="Activity type"
        valuesMapper={typeToText}
      />
      <SearchSelect
        name="userEmail"
        options={users}
        onChange={(value: string) => setValue('userEmail', value)}
        placeholder="User email"
        error={get(errors.userEmail, 'message')}
        value={watch('userEmail')}
      />
      <Calendar
        onChangeFromDateInput={(value: Moment) => setValue('fromDate', value)}
        onChangeToDateInput={(value: Moment) => setValue('toDate', value)}
        formFromDate={watch('fromDate')}
        formToDate={watch('toDate')}
        error={get(errors.fromDate, 'message') || get(errors.toDate, 'message')}
        hideError
      />
      <div className={styles.buttons}>
        <Button
          label={'SEARCH'}
          onClick={handleSubmit(handleOnSubmit)}
          border
          style={{ margin: '24px 0' }}
        />
        <Button label={'CLEAR'} onClick={reset} style={{ margin: '24px 0' }} />
      </div>
    </form>
  );
}

export default FiltersBar;
