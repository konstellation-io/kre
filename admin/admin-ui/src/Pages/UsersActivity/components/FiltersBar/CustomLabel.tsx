import React, { FC } from 'react';
import cx from 'classnames';
import styles from './FiltersBar.module.scss';

export enum HIGHLIGHT_COLORS {
  DEFAULT = 'highlight',
  INFO = 'feedback',
  SUCCESS = 'success',
  GREY = 'regular'
}
type Props = {
  children: string;
  highlight?: string;
  color?: HIGHLIGHT_COLORS;
};
const CustomLabel: FC<Props> = ({ children, highlight, color }) => {
  const highlightedText = highlight;
  const normalText = children.replace(highlight || '', '');

  return (
    <div className={styles.customLabel}>
      <span>{normalText}</span>
      {highlightedText && (
        <span className={cx(styles.customLabelHighlight, styles[color || ''])}>
          {highlightedText}
        </span>
      )}
    </div>
  );
};

export default CustomLabel;
