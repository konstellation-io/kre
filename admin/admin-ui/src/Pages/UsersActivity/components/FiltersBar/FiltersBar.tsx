import CustomLabel, { HIGHLIGHT_COLORS } from './CustomLabel';
import { FieldError, NestDataObject } from 'react-hook-form';
import GroupSelect, {
  GroupSelectData
} from 'Components/Form/GroupSelect/GroupSelect';

import ActivityIcon from '../ActivityIcon/ActivityIcon';
import Button from 'Components/Button/Button';
import Calendar from 'Components/Form/Calendar/Calendar';
import ErrorMessage from 'Components/ErrorMessage/ErrorMessage';
import { GetUsers } from 'Graphql/queries/types/GetUsers';
import { GetUsersActivity_userActivityList_user } from 'Graphql/queries/types/GetUsersActivity';
import { Moment } from 'moment';
import MultiSelect from 'Components/Form/MultiSelect/MultiSelect';
import React from 'react';
import SearchSelect from 'Components/Form/SearchSelect/SearchSelect';
import { UserActivityFormData } from '../../UsersActivity';
import { UserActivityType } from 'Graphql/types/globalTypes';
import { VersionsData } from 'Hooks/useAllVersions';
import { get } from 'lodash';
import { loader } from 'graphql.macro';
import styles from './FiltersBar.module.scss';
import { useQuery } from '@apollo/react-hooks';

const GetUsersQuery = loader('Graphql/queries/getUsers.graphql');

const customLabels = new Map([
  [UserActivityType.LOGIN, <CustomLabel>Login</CustomLabel>],
  [UserActivityType.LOGOUT, <CustomLabel>Logout</CustomLabel>],
  [UserActivityType.CREATE_RUNTIME, <CustomLabel>Runtime created</CustomLabel>],
  [UserActivityType.CREATE_VERSION, <CustomLabel>Version Created</CustomLabel>],
  [
    UserActivityType.PUBLISH_VERSION,
    <CustomLabel highlight="published" color={HIGHLIGHT_COLORS.DEFAULT}>
      Version published
    </CustomLabel>
  ],
  [
    UserActivityType.UNPUBLISH_VERSION,
    <CustomLabel highlight="unpublished" color={HIGHLIGHT_COLORS.INFO}>
      Version unpublished
    </CustomLabel>
  ],
  [
    UserActivityType.STOP_VERSION,
    <CustomLabel highlight="stopped" color={HIGHLIGHT_COLORS.GREY}>
      Version stopped
    </CustomLabel>
  ],
  [
    UserActivityType.START_VERSION,
    <CustomLabel highlight="started" color={HIGHLIGHT_COLORS.SUCCESS}>
      Version started
    </CustomLabel>
  ],
  [
    UserActivityType.UPDATE_SETTING,
    <CustomLabel>Settings updated</CustomLabel>
  ],
  [
    UserActivityType.UPDATE_VERSION_CONFIGURATION,
    <CustomLabel>Version conf. changed</CustomLabel>
  ],
  [UserActivityType.CREATE_USER, <CustomLabel>User created</CustomLabel>],
  [UserActivityType.REMOVE_USERS, <CustomLabel>Users removed</CustomLabel>],
  [
    UserActivityType.UPDATE_ACCESS_LEVELS,
    <CustomLabel>Users updated</CustomLabel>
  ],
  [UserActivityType.REVOKE_SESSIONS, <CustomLabel>Session revoked</CustomLabel>]
]);

const multiSelectOptions = Object.values(UserActivityType).map(type => ({
  label: type,
  Icon: <ActivityIcon activityType={type} invert />
}));

type FormFieldProps = {
  setAndSubmit: (
    field: string,
    newValue: string | GroupSelectData | Moment | UserActivityType[]
  ) => void;
  runtimesAndVersions: VersionsData[];
  watch: Function;
  errors: NestDataObject<UserActivityFormData, FieldError>;
  reset: Function;
};
function FiltersBar({
  setAndSubmit,
  runtimesAndVersions,
  watch,
  errors,
  reset
}: FormFieldProps) {
  const { data: usersData, error: usersError } = useQuery<GetUsers>(
    GetUsersQuery
  );
  const users =
    (usersData &&
      usersData.users.map(
        (user: GetUsersActivity_userActivityList_user) => user.email
      )) ||
    [];

  if (usersError) return <ErrorMessage />;

  function onTypeSelection(newtypes: UserActivityType[]) {
    setAndSubmit('types', newtypes);
  }

  const versionOptions = Object.fromEntries(
    runtimesAndVersions.map(({ runtime: { name: runtimeName }, versions }) => [
      runtimeName,
      versions.map(v => v.name)
    ])
  );

  return (
    <form className={styles.formField}>
      <div>
        <SearchSelect
          name="userEmail"
          label="search a user"
          options={users}
          onChange={(value: string) => setAndSubmit('userEmail', value)}
          placeholder="User email"
          error={get(errors.userEmail, 'message') as string}
          value={watch('userEmail')}
        />
      </div>
      <div>
        <GroupSelect
          label="filter by versions"
          options={versionOptions}
          formSelectedOptions={watch('versionIds')}
          onChange={newSelection => setAndSubmit('versionIds', newSelection)}
          placeholder="Select Versions"
          hideSelections
        />
      </div>
      <div>
        <MultiSelect<UserActivityType>
          options={multiSelectOptions}
          onChange={onTypeSelection}
          error={get(errors.types, 'message')}
          formSelectedOptions={watch('types')}
          label="Activity type"
          placeholder="ALL TYPES"
          selectionUnit="TYPE"
          selectAllText="ALL TYPES"
          customLabels={customLabels}
          iconAtStart
        />
      </div>
      <div>
        <Calendar
          label="dates selection"
          onChangeFromDateInput={(value: Moment) =>
            setAndSubmit('fromDate', value)
          }
          onChangeToDateInput={(value: Moment) => setAndSubmit('toDate', value)}
          formFromDate={watch('fromDate')}
          formToDate={watch('toDate')}
          error={
            get(errors.fromDate, 'message') || get(errors.toDate, 'message')
          }
        />
      </div>
      <div className={styles.buttons}>
        <Button
          label={'CLEAR'}
          onClick={() => reset()}
          style={{ margin: '24px 0' }}
        />
      </div>
    </form>
  );
}

export default FiltersBar;
