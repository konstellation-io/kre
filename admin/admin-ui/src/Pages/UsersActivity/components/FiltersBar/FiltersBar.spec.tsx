import FiltersBar from './FiltersBar';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('@apollo/react-hooks', () => ({
  useQuery: jest.fn(() => ({}))
}));

describe('FiltersBar', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(
      <FiltersBar
        onSubmit={jest.fn()}
        types={['a', 'b', 'c']}
        users={['1', '2', '3']}
        runtimesAndVersions={[]}
        errors={{}}
        watch={jest.fn()}
      />
    );
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
