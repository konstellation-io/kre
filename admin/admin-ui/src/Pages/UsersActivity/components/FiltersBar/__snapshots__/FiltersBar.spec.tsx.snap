// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersBar matches snapshot 1`] = `
<form
  className="formField"
>
  <div>
    <SearchSelect
      label="search a user"
      name="userEmail"
      onChange={[Function]}
      options={Array []}
      placeholder="User email"
    />
  </div>
  <div>
    <GroupSelect
      hideSelections={true}
      label="filter by versions"
      onChange={[Function]}
      options={Object {}}
      placeholder="Select Versions"
    />
  </div>
  <div>
    <MultiSelect
      customLabels={
        Map {
          "LOGIN" => <CustomLabel>
            Login
          </CustomLabel>,
          "LOGOUT" => <CustomLabel>
            Logout
          </CustomLabel>,
          "CREATE_RUNTIME" => <CustomLabel>
            Runtime created
          </CustomLabel>,
          "CREATE_VERSION" => <CustomLabel>
            Version Created
          </CustomLabel>,
          "PUBLISH_VERSION" => <CustomLabel
            color="highlight"
            highlight="published"
          >
            Version published
          </CustomLabel>,
          "UNPUBLISH_VERSION" => <CustomLabel
            color="feedback"
            highlight="unpublished"
          >
            Version unpublished
          </CustomLabel>,
          "STOP_VERSION" => <CustomLabel
            color="regular"
            highlight="stopped"
          >
            Version stopped
          </CustomLabel>,
          "START_VERSION" => <CustomLabel
            color="success"
            highlight="started"
          >
            Version started
          </CustomLabel>,
          "UPDATE_SETTING" => <CustomLabel>
            Settings updated
          </CustomLabel>,
          "UPDATE_VERSION_CONFIGURATION" => <CustomLabel>
            Version conf. changed
          </CustomLabel>,
          "CREATE_USER" => <CustomLabel>
            User created
          </CustomLabel>,
          "REMOVE_USERS" => <CustomLabel>
            Users removed
          </CustomLabel>,
          "UPDATE_ACCESS_LEVELS" => <CustomLabel>
            Users updated
          </CustomLabel>,
          "REVOKE_SESSIONS" => <CustomLabel>
            Session revoked
          </CustomLabel>,
        }
      }
      iconAtStart={true}
      label="Activity type"
      onChange={[Function]}
      options={
        Array [
          Object {
            "Icon": <UNDEFINED
              activityType="CREATE_RUNTIME"
              invert={true}
            />,
            "label": "CREATE_RUNTIME",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="CREATE_USER"
              invert={true}
            />,
            "label": "CREATE_USER",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="CREATE_VERSION"
              invert={true}
            />,
            "label": "CREATE_VERSION",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="LOGIN"
              invert={true}
            />,
            "label": "LOGIN",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="LOGOUT"
              invert={true}
            />,
            "label": "LOGOUT",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="PUBLISH_VERSION"
              invert={true}
            />,
            "label": "PUBLISH_VERSION",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="REMOVE_USERS"
              invert={true}
            />,
            "label": "REMOVE_USERS",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="REVOKE_SESSIONS"
              invert={true}
            />,
            "label": "REVOKE_SESSIONS",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="START_VERSION"
              invert={true}
            />,
            "label": "START_VERSION",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="STOP_VERSION"
              invert={true}
            />,
            "label": "STOP_VERSION",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="UNPUBLISH_VERSION"
              invert={true}
            />,
            "label": "UNPUBLISH_VERSION",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="UPDATE_ACCESS_LEVELS"
              invert={true}
            />,
            "label": "UPDATE_ACCESS_LEVELS",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="UPDATE_SETTING"
              invert={true}
            />,
            "label": "UPDATE_SETTING",
          },
          Object {
            "Icon": <UNDEFINED
              activityType="UPDATE_VERSION_CONFIGURATION"
              invert={true}
            />,
            "label": "UPDATE_VERSION_CONFIGURATION",
          },
        ]
      }
      placeholder="ALL TYPES"
      selectAllText="ALL TYPES"
      selectionUnit="TYPE"
    />
  </div>
  <div>
    <Calendar
      label="dates selection"
      onChangeFromDateInput={[Function]}
      onChangeToDateInput={[Function]}
    />
  </div>
  <div
    className="buttons"
  >
    <Button
      label="CLEAR"
      onClick={[Function]}
      style={
        Object {
          "margin": "24px 0",
        }
      }
    />
  </div>
</form>
`;
